src/helpers/types.ts

```diff
@@ -0,0 +1,24 @@
+export type WithChildren<T> = T & { children?: React.ReactNode };
+
+export type WithRequiredChildren<T> = T & { children: React.ReactNode };
+
+export interface AsProp<C extends React.ElementType> {
+  as?: C;
+}
+
+export type PolyProps<C extends React.ElementType, P = unknown> = AsProp<C> &
+  Omit<React.ComponentPropsWithRef<C>, keyof AsProp<C>> &
+  P;
```

Comment:
This was taken from somewhere else. There's an alternative solution that doesn't rely on the complex typings - emotion does have a `SomeStyledParagraph.withComponent('label')` that will have the proper types for `<label>` elements, but it isn't super ergonomic to use.

The main thing I want is for TS to be happy if I choose to do `<StyledThing as="label" htmlFor="some_id" />` since `htmlFor` is a valid attribute for `label` but not for `div` or `p` or whatever.